gcc/rust/backend/rust-compile-pattern.cc

```diff
@@ -329,5 +335,134 @@ CompilePatternBindings::visit (HIR::StructPattern &pattern)
     }
 }
 
+void
+CompilePatternBindings::visit (HIR::GroupedPattern &pattern)
+{
+  pattern.get_item ()->accept_vis (*this);
+}
+
+void
+CompilePatternLet::visit (HIR::IdentifierPattern &pattern)
+{
+  Bvariable *var = nullptr;
+  if (!ctx->lookup_var_decl (pattern.get_pattern_mappings ().get_hirid (),
+			     &var))
+    {
+      // FIXME this should be an assertion instead and use error mark node
+      rust_fatal_error (pattern.get_locus (),
+			"failed to lookup compiled variable declaration");
+      return;
+    }
+
+  auto fnctx = ctx->peek_fn ();
+  if (ty->is_unit ())
+    {
+      ctx->add_statement (init_expr);
+
+      tree stmt_type = TyTyResolveCompile::compile (ctx, ty);
+
+      auto unit_type_init_expr
+	= ctx->get_backend ()->constructor_expression (stmt_type, false, {}, -1,
+						       rval_locus);
+      auto s = ctx->get_backend ()->init_statement (fnctx.fndecl, var,
+						    unit_type_init_expr);
+      ctx->add_statement (s);
+    }
+  else
+    {
+      auto s
+	= ctx->get_backend ()->init_statement (fnctx.fndecl, var, init_expr);
+      ctx->add_statement (s);
+    }
+}
+
+void
+CompilePatternLet::visit (HIR::GroupedPattern &pattern)
+{
+  CompilePatternLet::Compile (&*pattern.get_item (), init_expr, ty, rval_locus,
+			      ctx);
+}
+
+void
+CompilePatternLet::visit (HIR::WildcardPattern &pattern)
+{
+  Bvariable *var = nullptr;
+  if (!ctx->lookup_var_decl (pattern.get_pattern_mappings ().get_hirid (),
+			     &var))
+    {
+      // FIXME this should be an assertion instead and use error mark node
+      rust_fatal_error (pattern.get_locus (),
+			"failed to lookup compiled variable declaration");
+      return;
+    }
+
+  auto fnctx = ctx->peek_fn ();
+  if (ty->is_unit ())
+    {
+      ctx->add_statement (init_expr);
+
+      tree stmt_type = TyTyResolveCompile::compile (ctx, ty);
+
+      auto unit_type_init_expr
+	= ctx->get_backend ()->constructor_expression (stmt_type, false, {}, -1,
+						       rval_locus);
+      auto s = ctx->get_backend ()->init_statement (fnctx.fndecl, var,
+						    unit_type_init_expr);
+      ctx->add_statement (s);
+    }
+  else
+    {
+      auto s
+	= ctx->get_backend ()->init_statement (fnctx.fndecl, var, init_expr);
+      ctx->add_statement (s);
+    }
+}
+
+void
+CompilePatternLet::visit (HIR::TuplePattern &pattern)
+{
+  rust_assert (!ty->is_unit ());
+  rust_assert (pattern.has_tuple_pattern_items ());
+  switch (pattern.get_items ()->get_pattern_type ())
+    {
+    default:
+      gcc_unreachable ();
+    case HIR::TuplePatternItems::TuplePatternItemType::RANGED:
+      rust_fatal_error (pattern.get_locus (),
+			"ranged tuple pattern in let statement");
+      return;
+    case HIR::TuplePatternItems::TuplePatternItemType::MULTIPLE:
+      break;
+    }
```

Comment:
1. Won't this switch statement always evaluate to the `default` case?
2. Are `RANGED` pattersn in let pattern a programmer error or something we do not support yet? If it is the former, I would switch to a call to `rust_error`, and in the case of the latter, `rust_sorry_at`